### DISCUSSION TITLE
Commits signed with my noreply email are seen as "Unverified"

### DISCUSSION THREAD

**Discussion by rosmanov** [community/community#41199] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Bug

### Body

GitHub shows GPG-signed commits as "Unverified" with the following error:

> The email in this signature doesn’t match the committer email. 

I have the "Keep my email addresses private" checkbox turned on in https://github.com/settings/emails and the tip under it states:

> We’ll remove your public profile email and use **608192+rosmanov@users.noreply.github.com** when...

And the committer email exactly matches that email. Let's take [this commit](https://github.com/rosmanov/bee-host/commit/a114408ba4e77cb429676a747d20be33d89a1bea), for example:

```
$ git --no-pager show --pretty=fuller --show-signature a114408ba4e77cb429676a747d20be33d89a1bea | head -n 10
commit a114408ba4e77cb429676a747d20be33d89a1bea
gpg: Signature made Wed 05 Oct 2022 07:32:24 PM +04
gpg:                using EDDSA key D49983485B84E1A9A3E20157B71B9215E724E859
gpg: Good signature from "Ruslan Osmanov (Git) <rrosmanov@gmail.com>" [ultimate]
gpg:                 aka "Ruslan Osmanov (GitHub) <608192+rosmanov@users.noreply.github.com>" [ultimate]
Author:     Ruslan Osmanov <608192+rosmanov@users.noreply.github.com>
AuthorDate: Wed Oct 5 19:32:24 2022 +0400
Commit:     Ruslan Osmanov <608192+rosmanov@users.noreply.github.com>
CommitDate: Wed Oct 5 19:32:24 2022 +0400
```

But GitHub displays it as "Unverified":
![unverified-commit-github](https://user-images.githubusercontent.com/608192/206724212-12958422-a06c-4f4b-a3f5-f0744ee1e979.png)

It looks like a bug. If it is not, please help me to fix this issue.

Thanks.

**Comment by ilyar** [comment]
![image](https://user-images.githubusercontent.com/761285/225586273-251644e5-3a84-461e-81c8-181ed8f8d9e5.png)
https://github.com/ilyar/test-gpg/commit/4487ec7e82e739b7f67fff21a04b4c2aee8ccdaf
![image](https://user-images.githubusercontent.com/761285/225586642-b84071f8-e4c6-43ea-a3f4-f3a680389f1a.png)
https://github.com/ilyar/test-gpg/commits/main
![image](https://user-images.githubusercontent.com/761285/225590744-cbdbcbe4-27dc-4a8c-8985-6ecc3eb8e89b.png)

**Comment by ilyar** [comment]
```shell
$ git commit --help | grep sign
       -s, --signoff
           Add Signed-off-by line by the committer at the end of the commit log message. The meaning of a signoff
       -S[<keyid>], --gpg-sign[=<keyid>]
           GPG-sign commits. The keyid argument is optional and defaults to the committer identity; if specified,
       --no-gpg-sign
           signed.
       designed not to force UTF-8 on projects. If all participants of a particular project find it more convenient
```

```shell
$ git commit -s -m "change title"
$ git --no-pager show --pretty=fuller --show-signature 1c69f1800e6a4ca764d9d0b5027c2ba962cd5b5f
commit 1c69f1800e6a4ca764d9d0b5027c2ba962cd5b5f
Author:     Ilyar <761285+ilyar@users.noreply.github.com>
AuthorDate: Thu Mar 16 11:21:11 2023 +0100
Commit:     Ilyar <761285+ilyar@users.noreply.github.com>
CommitDate: Thu Mar 16 11:21:11 2023 +0100

    change title
    
    Signed-off-by: Ilyar <761285+ilyar@users.noreply.github.com>

```
https://github.com/ilyar/test-gpg/commit/1c69f1800e6a4ca764d9d0b5027c2ba962cd5b5f
![image](https://user-images.githubusercontent.com/761285/225599508-a523381e-6c15-4177-923d-8cde259156e2.png)

```shell
$ git commit -S -m "use key"
$ git --no-pager show --pretty=fuller --show-signature 4b1fa45ac86305fd955397d95694350fc700f670
commit 4b1fa45ac86305fd955397d95694350fc700f670 (HEAD -> main, origin/main, origin/HEAD)
gpg: Signature made Thu 16 Mar 2023 12:05:34 PM CET
gpg:                using RSA key 01C0F1BCFA480C23C24793E04331FA7A0AF8BF99
gpg: Good signature from "Ilyar <761285+ilyar@users.noreply.github.com>" [ultimate]
Author:     Ilyar <761285+ilyar@users.noreply.github.com>
AuthorDate: Thu Mar 16 12:05:34 2023 +0100
Commit:     Ilyar <761285+ilyar@users.noreply.github.com>
CommitDate: Thu Mar 16 12:05:34 2023 +0100
```
https://github.com/ilyar/test-gpg/commit/4b1fa45ac86305fd955397d95694350fc700f670
![image](https://user-images.githubusercontent.com/761285/225599643-910e0d95-e994-4c7c-bf2d-2956deb2cf11.png)

**Reply by nikicat** [reply]
I did all the same and it doesn't work - my commits are still "Unverified"

**Comment by lasersPew** [comment]
When you change your email to GitHub throwaway email, you should add new user id to the GPG key you use when signing commits to GitHub. More info: [StackOverflow](https://security.stackexchange.com/questions/261467/can-i-add-an-email-address-to-an-existing-gpg-key)

**Comment by asamahy** [comment]
after adding the new ID to the key, remove the old public key from github and add the updated key. your old commits wont be affected.
`gpg --armor --export your-key-id`

**Comment by jimmydju** [comment]
You need to add your email `608192+rosmanov@users.noreply.github.com`  to your local GPG key (follow [this](https://security.stackexchange.com/questions/261467/can-i-add-an-email-address-to-an-existing-gpg-key)) and then re-generate the public key. 
Go to github, delete the existing public key and add the newly generated public key. Your next commit/push should be verified by Github now.

**Reply by egouldo** [reply]
I generated a gpg key for the first time, doing exactly this, and I'm still getting this problem, not a solution to the OP's problem unfortunately.

**Comment by idontlikecreatingaccounts** [comment]
FIXED: I had an extra whitespace in the git config email address.

**Comment by nikicat** [comment]
The same issue for me, @rosmanov did you manage to find a solution?

**Comment by nikicat** [comment]
I've found a solution - you need to remove a digit prefix from the `noreply` email address (just leave a GitHub username)

**Comment by Kristinita** [comment]
**Type: Question** :question:

@nikicat, could you provide more details? Where should the user remove the digit prefix?

### 1. Local git configuration

Should the user replace the email in the local git configuration? Example:

```sh
# [INFO] Get current email
git config --global user.email
17247677+Kristinita@users.noreply.github.com

# [QUESTION] Should the email be like this?
git config --global user.email
Kristinita@users.noreply.github.com
```

### 2. E-mail for GPG key

Or perhaps the user should add the GitHub no-reply e-mail to the GPG key without the digit prefix? For example, my page **<https://github.com/settings/keys>**:

![GitHub GPG keys](https://i.imgur.com/EWQTMuf.png)

Should the email address for GPG key not contain `17247677+` in my case?

---

Or did you mean something else?

Thanks.

**Reply by nikicat** [reply]
You are correct - I've removed the digit prefix from my private GitHub-generated email address in commits and they became "Verified".

**Reply by Kristinita** [reply]
**Status: NOT REPRODUCED** 🔁

Unfortunately, it doesn’t work for me.

![GitHub no verify e-mail without a digit prefix](https://i.imgur.com/CefKjRS.png)

### 1. Steps to reproduce

> You are correct - I've removed the digit prefix from my private GitHub-generated email address in commits and they became "Verified".

If I understood “in commits” correctly, I should remove a digit prefix in e-mail in my global and repository git settings.

```text
git config --global user.email
Kristinita@users.noreply.github.com

git config user.email
Kristinita@users.noreply.github.com
```

I make test changes in my demo repository → `git add .` → `git commit -m "Commit description"` → `git push origin HEAD`:

```text
git log --show-signature -1
commit a735fc976b535938cb6fb4bbd2cee463ca2c8407 (HEAD -> KiraVerifyCommitWithNoReplyEMail, origin/KiraVerifyCommitWithNoReplyEMail)
gpg: Signature made 1/15/2024 2:47:32 PM Russia TZ 2 Standard Time^M
gpg:                using RSA key <my real 40-digit RSA key>^M
gpg: Good signature from "Kristinita (GitHub no-reply email address) <17247677+Kristinita@users.noreply.github.com>" [ultimate]^M
gpg:                 aka "Kristinita <myrealemail@kiragoddess.com>" [ultimate]^M
Author: Kristinita <Kristinita@users.noreply.github.com>
Date:   Mon Jan 15 14:47:32 2024 +0300

	vcs(git): try remove a digit prefix from GitHub no-reply e-mail address in local configuration
```

### 2. Current behavior

[**Still `Unverified`**](https://github.com/Kristinita/SashaGitHub/commit/a735fc976b535938cb6fb4bbd2cee463ca2c8407):

> The email in this signature doesn’t match the committer email.

![Commit is still unverified](https://i.imgur.com/o0IMu6X.png)

Thanks.

**Reply by nikicat** [reply]
I think you should also change the email address in the signature.

**Reply by Kristinita** [reply]
**Type: Question** :question:

@nikicat, could you, in the repository, where GitHub successfully verifies your commits, run the command locally?

```text
git log --show-signature -1
```

And paste the output here, hiding your real e-mail and RSA key, as in my example.

Thanks.

**Reply by nikicat** [reply]
Sure, there is no public email in this commit btw
```
commit 7ea0161dc13754c9c241bdbdebab1b2ec995d34d (HEAD -> journald, origin/journald)
gpg: Signature made Tue Jan  9 11:23:21 2024 UTC
gpg:                using EDDSA key B374CC3950AB7E4CDAAC6986846FFFEFC1039264
gpg: Good signature from "Nikolay Bryskin (GitHub) <nikicat@users.noreply.github.com>" [unknown]
gpg:                 aka "Nikolay Bryskin (GitHub) <329262+nikicat@users.noreply.github.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: B374 CC39 50AB 7E4C DAAC  6986 846F FFEF C103 9264
Author: Nikolay Bryskin <nikicat@users.noreply.github.com>
Date:   Tue Jan 9 10:20:14 2024 +0000

    Add CLA
    
    Signed-off-by: Nikolay Bryskin <nikicat@users.noreply.github.com>
```

**Reply by Kristinita** [reply]
But the fact that GitHub e-mail with ID number is `Unverified` looks like a GitHub flaw.

![GitHub e-mail Unverified and Verified](https://i.imgur.com/6svALVn.png)

**Reply by Kristinita** [reply]
**Status: RESOLVED** :heavy_check_mark:

[**Thanks**](https://github.com/Kristinita/SashaGitHub/commit/619fcb1ebff602670faf5f5d8557191ef633b269)!

![Verified commit with GitHub e-mail without ID number](https://i.imgur.com/5PNbpCY.png)

**Reply by nikicat** [reply]
I agree, it should be fixed.

**Reply by Aurora12** [reply]
The problem still persists. 🙄
I've just lost a lot of time on it — thanks for the solution!
<img width="440" height="192" alt="image" src="https://github.com/user-attachments/assets/965cb057-3ac2-47a2-b553-c25dd4ffac0f" />

And it looks like I need `Aurora12@users.noreply.github.com` in `git config` too. They can't be different. In which case, will everything else on GitHub work, I wonder? 🤔 I mean, GitHub profile attribution, contribution graph, etc...

**Comment by wirredfunds** [comment]
**I fixed this issue!**

After two long days of searching, I have finally found out how to fix the issue

# The Problem

No, the problem wasn't because the email wasn't in the key itself nor was it because I didn't put it to GitHub.

The problem is that by default, **GPG keypairs don't have signing privileges on the main key**. This is a *security practice* done by GnuPG to ensure you can't sign everything with your main key and need to share a *sub*key.

## The Solution

1. Run `gpg --list-keys --keyid-format=long`
2. Find your key and don't use the **pub ID**, use the ***key ID you found on the sub*** section, also known as the subkey section which is used for signing by default.
3. Run `gpg --export --armor [subKeyID]` and put the *subkey* into GitHub's GPG Keys and not the public key itself.
4. Put the new signing key for Git usage: `git config user.signingkey [subKeyID]` and use `git config commit.gpgsign true` to automatically sign all of your commits.

### If the solution doesn't work...

*I have no hope for you...*
